### DISCUSSION TITLE
ch4/ch4-03

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#157] [General]
# ch4/ch4-03

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch4/ch4-03.html

**Comment by zhangbest5** [comment]
ages["bob"] = ages["bob"] + 1 // happy birthday!
这个注释让我很费解，这是结果？还是想表达啥？

**Reply by zhangbest5** [reply]
结果是增加了 bob:1 这个键值对，这注释根本无意义吧？

**Reply by FanXieGangN** [reply]
这里就是抖了个机灵，bob本来不在ages里面，所以ages["bob"]是0，现在+1变成了1岁，相当于刚出生，就祝他生日快乐。

**Reply by zhangbest5** [reply]
😄，感谢大佬回复，理解了

**Comment by zhangbest5** [comment]
gopl.io/ch4/graph 突然抛出个2维Map表示很懵逼

**Comment by xianren68** [comment]
### 4.8
```
func main() {
	charCount()
}
func charCount() {
	// 创建hash表
	hash := make(map[string]int, 4)
	input := bufio.NewReader(os.Stdin)
	for {
		r, _, err := input.ReadRune()
		if err == io.EOF { // 输入ctrl+z
			break
		}
		if err != nil {
			fmt.Fprintf(os.Stderr, "charcount: %v\n", err)
			os.Exit(1)
		}
		t := whatType(r)
		hash[t]++
	}
	fmt.Printf("%#v", hash)
}

// 判断一个unicode码点为什么类型
func whatType(r rune) string {
	if unicode.IsNumber(r) {
		return "number"
	}
	if unicode.Is(unicode.Han, r) {
		return "中文"
	}
	if r >= 65 && r <= 122 {
		return "English"
	}

	return "any"
}
```

**Comment by xianren68** [comment]
### 4.9
```
// 统计词频
func wordfreq() {
	hash := make(map[string]int, 20)
	input := bufio.NewScanner(os.Stdin)
	input.Split(bufio.ScanWords)
	for input.Scan() {
		hash[input.Text()]++
	}
	fmt.Printf("%#v", hash)
}
```

**Comment by xq-120** [comment]
就不能多加个set类型，搞得这么麻烦

**Comment by SuperZ3** [comment]
4.9
```go
package main

import (
    "bufio"
    "fmt"
    _ "io"
    "os"
    _ "unicode"
    _ "unicode/utf8"
)

func wordfreq() {
    counter := make(map[string]int)
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Split(bufio.ScanWords)
    for scanner.Scan() {
        word := scanner.Text()
        counter[word]++
        fmt.Println(counter[word])
	}
    if err := scanner.Err(); err != nil {
		fmt.Printf("Invalid input: %s", err)
	}
}

func main() {
   wordfreq()
}
```

**Comment by nwnusun-cool** [comment]
equal(map[string]int{"A": 0}, map[string]int{"B": 42})
返回的肯定是false,所以不相等啊

**Reply by soondubu137** [reply]
如果直接用`xv != y[k]`判定的话就会返回true了。因为第二个map里不存在“A”，所以查找“A”的话会返回0，这个因为不存在而返回的0会错误地和第一个map里真正存在的0被判定为一致

**Comment by eroneko** [comment]
要是from这个key不存在，hasEdge不会panic吗？不是很懂

**Reply by liushunkkk** [reply]
nil 的 map 是可读的，但是不可赋值，不可修改

**Comment by leolin49** [comment]
> 禁止对map元素取址的原因是map可能随着元素数量的增长而重新分配更大的内存空间，从而可能导致之前的地址无效。
那为什么可以对slice中的元素取地址，slice不是也有可能会 “随着元素数量的增长而重新分配更大的内存空间，从而可能导致之前的地址无效” 吗？

**Comment by hongjiala** [comment]
// 练习 4.8： 修改charcount程序，使用unicode.IsLetter等相关的函数，统计字母、数字等Unicode中不同的字符类别。
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"unicode"
)

func main() {
	counts := make(map[string]int) // counts of Unicode characters

	in := bufio.NewReader(os.Stdin)
	for {
		r, n, err := in.ReadRune() // returns rune, nbytes, error
		if err == io.EOF {
			break
		}
		if err != nil {
			fmt.Fprintf(os.Stderr, "charcount: %v\n", err)
			os.Exit(1)
		}
		if r == unicode.ReplacementChar && n == 1 {
			counts["invalid"]++
			continue
		}
		if unicode.IsLetter(r) {
			counts["letter"]++
			continue
		}
		if unicode.IsNumber(r) {
			counts["number"]++
			continue
		}
		if unicode.IsSpace(r) {
			counts["space"]++
			continue
		}

	}
	fmt.Printf("type\tcount\n")
	for c, n := range counts {
		fmt.Printf("%q\t%d\n", c, n)
	}

}

**Comment by hongjiala** [comment]
// 练习 4.9： 编写一个程序wordfreq程序，报告输入文本中每个单词出现的频率。在第一次调用Scan前先调用input.Split(bufio.ScanWords)函数，这样可以按单词而不是按行输入。
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"unicode"
)

func main() {
	freq := wordfreq(os.Stdin)
	for k, v := range freq {
		fmt.Printf("%s\t%d\n", k, v)
	}

}

func wordfreq(r io.Reader) map[string]int {
	input := bufio.NewScanner(r)
	input.Split(bufio.ScanWords)
	freq := make(map[string]int)
	for input.Scan() {
		word := input.Text()
		if unicode.IsLetter(rune(word[0])) {
			freq[word]++
		}
	}
	return freq
}

**Comment by celiachu** [comment]
### 4.8
```go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main(){
	counts := make(map[string]int)
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	for scanner.Scan() {
		word := scanner.Text()
		counts[word] ++
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "scanner error: ", err)
		os.Exit(1)
	}
	for word, count := range counts {
		fmt.Printf("%s: %d\n", word, count)
	}
}
```

**Comment by w2807** [comment]
4.9
```go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	mp := make(map[string]int)
	input := bufio.NewScanner(os.Stdin)
	input.Split(bufio.ScanWords)
	for input.Scan() {
		word := input.Text()
		mp[word]++
	}
	for i, j := range mp {
		fmt.Printf("%s: %d\n", i, j)
	}
}
```
